ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java

```diff
@@ -448,49 +452,108 @@ private String formatExpression(final Expression expression) {
+        final List<SqlArgument> argForReturnType = new ArrayList<>();
+        for (int i = 0; i < arguments.size(); i++) {
+          final Expression expression = arguments.get(i);
+          final ParamType parameter = paramTypes.get(i);
+          if (expression instanceof LambdaFunctionCall) {
+            final TypeContext childContext = context.getCopy();
+
+            // the function returned from the UDF factory should have lambda
+            // at this index in the function arguments if there's a
+            // lambda node at this index in the function node argument list
+            if (!(parameter instanceof LambdaType)) {
+              throw new KsqlException("Error while processing lambda function. "
+                  + "This is most likely an internal error and a "
+                  + "Github issue should be filed for debugging.");
+            }
+
+            final LambdaType lambdaParameter = (LambdaType) parameter;
+            for (ParamType inputParam : lambdaParameter.inputTypes()) {
+              if (inputParam instanceof GenericType) {
+                final GenericType genericParam = (GenericType) inputParam;
+                if (!reservedGenerics.containsKey(genericParam)) {
+                  throw new KsqlException(
+                      String.format(
+                          "Could not verify type for generic %s.",
+                          genericParam.toString()));
+                }
+                childContext.addLambdaInputType(reservedGenerics.get(genericParam));
+              } else {
+                childContext.addLambdaInputType(
+                    SchemaConverters.functionToSqlConverter().toSqlType(inputParam));
+              }
+            }
+            final List<SqlType> sqlTypesForLambdaArgMapping =
+                new ArrayList<>(childContext.getLambdaInputTypes());
+            final SqlType resolvedArgType = expressionTypeManager.getExpressionSqlType(
+                expression,
+                childContext.getCopy()
+            );
+            final SqlArgument lambdaArgument = SqlArgument.of(
+                SqlLambda.of(sqlTypesForLambdaArgMapping, resolvedArgType));
+
+            argForReturnType.add(lambdaArgument);
+            typeContextsForChildren.add(childContext);
+
+            GenericsUtil.reserveGenerics(
+                parameter,
+                lambdaArgument,
+                reservedGenerics
+            );
+          } else {
+            argForReturnType.add(argTypesForFirstPass.get(i));
+            typeContextsForChildren.add(context.getCopy());
+
+            GenericsUtil.reserveGenerics(
+                parameter,
+                argTypesForFirstPass.get(i),
+                reservedGenerics
+            );
+          }
+        }
+        functionReturnSchema = function.getReturnType(argForReturnType);
+        argumentSchemas.addAll(argForReturnType);
+      } else {
+        functionReturnSchema = function.getReturnType(argTypesForFirstPass);
```

Comment:
If i'm reading everything correctly, getting the return type is the one line different between here and `CodeGenRunner` ): IT looks like if we did pull this out into something like `genericsUtil` it would require a huge argument list... but it would feel nice to reuse some of the code at least between this and `CodeGenRunner`. At the very least WDYT about pulling out the stuff inside of `if(node.hasLambdaFunctionCall())` into a helper function? Just thinking that if other people have to work on this and don't want to get bogged down by lambdas that could help readability